File: backend/app/models/task_log.py
"""
TaskLog model: Structured observability logs for task execution.
"""

import enum
import uuid
from datetime import datetime
from sqlalchemy import String, Integer, Text, DateTime, ForeignKey, Enum, Uuid
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy.sql import func
from sqlalchemy.dialects.postgresql import JSONB

from app.core.database import Base


class LogLevel(str, enum.Enum):
    """Log message severity levels."""
    DEBUG = "debug"
    INFO = "info"
    WARNING = "warning"
    ERROR = "error"


class TaskLog(Base):
    """
    Structured log entry for task execution observability.
    
    Provides stage-level logging with structured metadata for debugging,
    monitoring, and audit trails.
    
    Attributes:
        log_id: Auto-incrementing primary key (for performance)
        task_id: Foreign key to parent task
        stage: Pipeline stage that generated this log
        level: Log severity (debug/info/warning/error)
        message: Human-readable log message
        data_json: Structured metadata (JSONB)
        timestamp: When log was created
    """
    
    __tablename__ = "task_logs"
    
    # Primary key (serial integer for performance)
    log_id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        autoincrement=True
    )
    
    # Foreign key to task
    task_id: Mapped[uuid.UUID] = mapped_column(
        Uuid,
        ForeignKey("tasks.task_id", ondelete="CASCADE"),
        nullable=False,
        index=True,
        comment="Parent task ID"
    )
    
    # Log metadata
    stage: Mapped[str | None] = mapped_column(
        String(50),
        nullable=True,
        comment="Pipeline stage name"
    )
    
    level: Mapped[LogLevel] = mapped_column(
        Enum(LogLevel, name="log_level", native_enum=False),
        nullable=False,
        default=LogLevel.INFO,
        comment="Log severity level"
    )
    
    message: Mapped[str] = mapped_column(
        Text,
        nullable=False,
        comment="Human-readable log message"
    )
    
    data_json: Mapped[dict | None] = mapped_column(
        JSONB,
        nullable=True,
        comment="Structured metadata"
    )
    
    # Timestamp
    timestamp: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        nullable=False,
        server_default=func.now(),
        index=True,
        comment="When log was created"
    )
    
    # Relationships
    task: Mapped["Task"] = relationship(
        "Task",
        back_populates="logs"
    )
    
    def __repr__(self) -> str:
        return f"<TaskLog(log_id={self.log_id}, level={self.level.value}, stage={self.stage})>"

File: frontend/src/types/task-log.ts
/**
 * TaskLog Type Definitions
 * 
 * Represents audit trail of task processing events.
 */

export type LogLevel = 'INFO' | 'WARNING' | 'ERROR';

export interface TaskLog {
  log_id: string;                    // Unique identifier
  task_id: string;                   // Parent task (foreign key)
  log_level: LogLevel;               // Severity level
  message: string;                   // Log message
  timestamp: string;                 // ISO 8601 timestamp
  metadata?: Record<string, unknown>; // Optional structured data
}

/**
 * Create a new task log entry
 */
export function createTaskLog(
  taskId: string,
  level: LogLevel,
  message: string,
  metadata?: Record<string, unknown>
): TaskLog {
  return {
    log_id: crypto.randomUUID(),
    task_id: taskId,
    log_level: level,
    message,
    timestamp: new Date().toISOString(),
    metadata,
  };
}

/**
 * Get color for log level
 */
export function getLogLevelColor(level: LogLevel): string {
  switch (level) {
    case 'INFO':
      return 'text-gray-700';
    case 'WARNING':
      return 'text-yellow-600';
    case 'ERROR':
      return 'text-red-600';
    default:
      return 'text-gray-700';
  }
}
